connect/transforms/src/test/java/org/apache/kafka/connect/transforms/HeaderFromTest.java

```diff
@@ -0,0 +1,357 @@
+                });
+        }
+        return result;
+    }
+
+    private final HeaderFrom<SourceRecord> xform;
+
+    private final RecordBuilder originalRecordBuilder;
+    private final RecordBuilder expectedRecordBuilder;
+    private final List<String> transformFields;
+    private final List<String> headers;
+    private final HeaderFrom.Operation operation;
+
+    public HeaderFromTest(String description,
+                          boolean keyTransform,
+                          RecordBuilder originalBuilder,
+                          List<String> transformFields, List<String> headers, HeaderFrom.Operation operation,
+                          RecordBuilder expectedBuilder) {
+        this.keyTransform = keyTransform;
+        this.xform = keyTransform ? new HeaderFrom.Key<>() : new HeaderFrom.Value<>();
+        this.originalRecordBuilder = originalBuilder;
+        this.expectedRecordBuilder = expectedBuilder;
+        this.transformFields = transformFields;
+        this.headers = headers;
+        this.operation = operation;
+    }
+
+    private Map<String, Object> config() {
+        Map<String, Object> result = new HashMap<>();
+        result.put(HeaderFrom.HEADERS_FIELD, headers);
+        result.put(HeaderFrom.FIELDS_FIELD, transformFields);
+        result.put(HeaderFrom.OPERATION_FIELD, operation.toString());
+        return result;
+    }
+
+    @Test
+    public void schemaless() {
+        xform.configure(config());
+        ConnectHeaders headers = new ConnectHeaders();
+        headers.addString("existing", "existing-value");
+
+        SourceRecord originalRecord = originalRecordBuilder.schemaless(keyTransform);
+        SourceRecord expectedRecord = expectedRecordBuilder.schemaless(keyTransform);
+        SourceRecord xformed = xform.apply(originalRecord);
+        assertSameRecord(expectedRecord, xformed);
+    }
+
+    @Test
+    public void withSchema() {
+        xform.configure(config());
+        ConnectHeaders headers = new ConnectHeaders();
+        headers.addString("existing", "existing-value");
+        Headers expect = headers.duplicate();
+        for (int i = 0; i < this.headers.size(); i++) {
+            expect.add(this.headers.get(i), originalRecordBuilder.fieldValues.get(i), originalRecordBuilder.fieldSchemas.get(i));
+        }
+
+        SourceRecord originalRecord = originalRecordBuilder.withSchema(keyTransform);
+        SourceRecord expectedRecord = expectedRecordBuilder.withSchema(keyTransform);
+        SourceRecord xformed = xform.apply(originalRecord);
+        assertSameRecord(expectedRecord, xformed);
+    }
+
+    @Test(expected = ConfigException.class)
+    public void invalidConfig() {
+        Map<String, Object> config = config();
+        List<String> headers = new ArrayList<>(this.headers);
+        headers.add("unexpected");
+        config.put(HeaderFrom.HEADERS_FIELD, headers);
+        xform.configure(config);
+    }
```

Comment:
Little silly to run this for every iteration of the parameterized test 😛 
I'm guessing there isn't an easy way to run this only once?

Reply:
Yeah, parameterized tests do have their faults. AFAIK the only way to handle this using Junit 4 is to have two tests. You can arrange to have both tests in the same source file so both parameterised and non-parameterised tests are still easily discoverable (e.g. [using Enclosed](https://stackoverflow.com/questions/3361239/excluding-a-non-param-test-in-parameterized-test-class)), but I've never seen that done in practice and it looks rather fiddly. Since these tests execute pretty quickly, I'm inclined to think we should just live with it. The real solution might just be to adopt Junit 5.